### PR TITLE
Use the conf of shuffleNodesNumber from jobs to be as checking factor

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?
Use the conf of shuffleNodesNumber from jobs to be as checking factor

### Why are the changes needed?
In the PR #97 , it allow client to specify the shuffle server numbers, but in clusterLoaderChecker, it dont take this into considering. In this PR, it will use the conf of shuffleNodesNumber from jobs to be as checking factor only when the conf of `rss.coordinator.access.loadChecker.serverNum.threshold` is missed.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
UTs.

### CODE REVIEW

**Comment by zuston** [comment]
PTAL @jerqi

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java [hunk 1]
```diff
@@ -99,13 +103,17 @@ private boolean tryAccessCluster() {
     long retryInterval = sparkConf.get(RssSparkConfig.RSS_CLIENT_ACCESS_RETRY_INTERVAL_MS);
     int retryTimes = sparkConf.get(RssSparkConfig.RSS_CLIENT_ACCESS_RETRY_TIMES);
 
+    int assignmentShuffleNodesNum = sparkConf.get(RssSparkConfig.RSS_CLIENT_ASSIGNMENT_SHUFFLE_SERVER_NUMBER);
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:106]: Why do we only modify spark3?

File: common/src/main/java/org/apache/uniffle/common/util/Constants.java [hunk 2]
```diff
@@ -43,4 +43,6 @@ public class Constants {
   public static final String CONF_REMOTE_STORAGE_PATH = ".remote.storage.path";
   public static final String RSS_CLIENT_CONF_REMOTE_STORAGE_PATH =
           RSS_CLIENT_CONF_COMMON_PREFIX + CONF_REMOTE_STORAGE_PATH;
+
+  public static final String ACCESS_INFO_REQUIRED_SHUFFLE_NODES_NUM = "access_info_required_shuffle_nodes_num";
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/util/Constants.java:47]: Why do we need this config option?

Comment (zuston) [reply common/src/main/java/org/apache/uniffle/common/util/Constants.java:47]: The key is the indicator of the required shuffle node number in the access info

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java (hunk 1 above)
Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:106]: Sorry, I forgot to modify in spark2

**Comment by jerqi** [comment]
`RSS_CLIENT_ASSIGNMENT_SHUFFLE_SERVER_NUMBER` means that the max number should be allocated for client. It's not min number that should be allocated for client. I think it's not suitable to judge whether to reject the application.

**Comment by zuston** [comment]
The conf of `rss.coordinator.shuffle.nodes.max` in coordinator is a soft limitation. It means when available nodes number is lower than this value, it will not throw exception and reassign the number of all available nodes.

But in current implementation of `AccessClusterLoadChecker`, it will reject the app when available nodes number < `rss.coordinator.shuffle.nodes.max`. 

### For example
available shuffle nodes number = 20  
job required shuffle nodes number = 10
`rss.coordinator.shuffle.nodes.max` = 30

The `AccessClusterLoadChecker` will reject this job. I think this is unreasonable.

**Comment by jerqi** [comment]
> The conf of `rss.coordinator.shuffle.nodes.max` in coordinator is a soft limitation. It means when available nodes number is lower than this value, it will not throw exception and reassign the number of all available nodes.
> 
> But in current implementation of `AccessClusterLoadChecker`, it will reject the app when available nodes number < `rss.coordinator.shuffle.nodes.max`.
> 
> ### For example
> available shuffle nodes number = 20 job required shuffle nodes number = 10 `rss.coordinator.shuffle.nodes.max` = 30
> 
> The `AccessClusterLoadChecker` will reject this job. I think this is unreasonable.

I got it.

**Comment by jerqi** [comment]
But the `rss.coordinator.shuffle.nodes.max` is just the default value. If you have other need, you can change this default value. `rss.coordinator.access.loadChecker.serverNum.threshold` isn't related the to `rss.coordinator.shuffle.nodes.max` actually.

**Comment by zuston** [comment]
You mean we should remove the check of `rss.coordinator.shuffle.nodes.max` in `AccessClusterLoadChecker`?

**Comment by jerqi** [comment]
> You mean we should remove the check of `rss.coordinator.shuffle.nodes.max` in `AccessClusterLoadChecker`?

Origin logic don't check `rss.coordinator.shuffle.nodes.max`. It's a config option `rss.coordinator.access.loadChecker.serverNum.threshold`. Its default value is `rss.coordinator.shuffle.nodes.max`. Default value don't need to be reasonable for every situation.

**Comment by zuston** [comment]
Got your thought. 

But the conf of `rss.coordinator.access.loadChecker.serverNum.threshold` is a hard limitation, if value is too low, it will pass all. If too high, it will reject all. 

For the equality of every job, I wont set it and hope the checker can decide whether to reject depending on the required shuffle servers number of jobs.

I thinks this is reasonable, it will still reserve the hard limitation and keep the flexibility

**Comment by jerqi** [comment]
> Got your thought.
> 
> But the conf of `rss.coordinator.access.loadChecker.serverNum.threshold` is a hard limitation, if value is too low, it will pass all. If too high, it will reject all.
> 
> For the equality of every job, I wont set it and hope the checker can decide whether to reject depending on the required shuffle servers number of jobs.
> 
> I thinks this is reasonable, it will still reserve the hard limitation and keep the flexibility

Should we give a option of coordinator to decide whether to use the required shuffle server of client for users.

**Comment by zuston** [comment]
> Should we give a option of coordinator to decide whether to use the required shuffle server of client for users?

Not necessary. If the conf value of `rss.client.assignment.shuffle.nodes.max` is -1 or not set, it will use the conf from coordinator side.

**Comment by jerqi** [comment]
> > Should we give a option of coordinator to decide whether to use the required shuffle server of client for users?
> 
> Not necessary. If the conf value of `rss.client.assignment.shuffle.nodes.max` is -1 or not set, it will use the conf from coordinator side.

If I want to use `rss.coordinator.access.loadChecker.serverNum.threshold`, it's weird that we must guarantee that the conf value of `rss.client.assignment.shuffle.nodes.max` is -1 or not set. They are not related config options.

**Comment by jerqi** [comment]
Like discussion at dev mail list, we will freeze the code and cut 0.6 version branch in September 15, we  will not merge this pr before I cut 0.6 version branch, are you ok?

**Comment by zuston** [comment]
> Like discussion at dev mail list, we will freeze the code and cut 0.6 version branch in September 15, we will not merge this pr before I cut 0.6 version branch, are you ok?

Got it

**Comment by zuston** [comment]
Ping @jerqi

**Comment by jerqi** [comment]
Should we add more documents for this feature?

**Comment by zuston** [comment]
> Should we add more documents for this feature?

Added.

**Review by jerqi — APPROVED** [review]
LGTM, thanks @zuston

**Comment by zuston** [comment]
Thanks for your patient review @jerqi

**Pull request merged**